### PR TITLE
Cc/feature/question pattern

### LINKED ISSUE #378: Add Question Form Pattern to Design System.
>**What does it do**
>
> - Asks a user questions which a service requires.
> - Helps to break down complex forms by spreading questions across multiple pages.
>
>**When to use it?**
>
> - When you are asking users for information as part of a service.
> - When a form is required to be spread across multiple pages.
>
>**When not to use it?**
>
> - When there are multiple questions on the same page, unless you have user research that says it makes sense to group the questions.
>
>**How it works**
>
> - Always test questions with users through user research.
> - Make sure your question makes sense.
> - Sometimes it makes sense to group a few questions on the same page.
> - Make sure users know why you're asking the question and only ask users for the information you really need.
> - Add (optional) to the end of each question that is optional.

> Add a new H2 heading named 'Form Element'.
>
> Add under the H2 heading: 
>
> This pattern can be populated with any form element, such as text, radio buttons, checkboxes, dropdown or file upload.
>
> _This is where the Questions Form / Form Element pattern will be placed._

> Add a new H2 heading named 'Reference'.
>
> Add under the H2 heading: 
>
> This pattern shows a question that shows defined steps and uses the backlink and progress indicator component.
>
> _This is where the Questions Forms / Reference pattern will be placed._

> Add a new H2 heading named 'Summary'.
>
> Add under the H2 heading: 
>
> This shows the user all of the data they have entered before they submit the form. This stage allows the user to change any details before submitting the form.
>
> _This is where the Questions Form / Summary pattern will be placed._

> Add a new H2 heading named 'Confirmation Message'.
>
> Add under the H2 heading: 
>
> This is a confirmation message to show that the user has successfully submitted the data in the form, along with a  reference or any information which the user needs to be made aware of.
>
> _This is where the Questions Form / Confirmation pattern will be placed._

### PR DESCRIPTION
**Pull request by catiarodriguescosta**
Fixes #378

### CODE REVIEW

File: src/wmnds/components/form-elements/input/_input.njk [hunk 1]
```diff
@@ -21,7 +24,7 @@
       wmndsFeErrorMessage() | trim
     }}
   {% endif -%}
-  <input class="wmnds-fe-input{{errorClass}}" id="{{id}}" name="{{id}}" type="text"/>
+  <input class="wmnds-fe-input{{errorClass}}{{paramClass}}" id="{{id}}" name="{{id}}" type="{{type}}"/>
```
[the remark this answers is not in the document]
Comment (catiarodriguescosta) [reply src/wmnds/components/form-elements/input/_input.njk:27]: It's not missing. Right before this, we have a label with the "for" linking it. 👍

[the remark this answers is not in the document]
Comment (daylesalmon) [reply src/wmnds/components/form-elements/input/_input.njk:27]: @catiarodriguescosta I think this is a none-issue as accesslint may not be able to understand nunjucks variables.

File: src/wmnds/assets/sass/_utilities.scss [hunk 2]
```diff
@@ -175,3 +190,23 @@ body {
...
+  }
+
+  &-gray {
+    background-color: get-color(secondary);
+  }
+
+  &-transparent {
+    background-color: transparent;
+  }
+}
```
Comment (daylesalmon) [inline src/wmnds/assets/sass/_utilities.scss:212]: Same with these, can you add them to utilities page.

File: src/wmnds/components/form-elements/input/_input.njk [hunk 3]
```diff
@@ -21,7 +24,7 @@
...
 </div>
 
 {% endmacro %}]
```
Comment (daylesalmon) [inline src/wmnds/components/form-elements/input/_input.njk:30]: "no newline at end of file"

File: src/wmnds/components/form-elements/input/_input.scss [hunk 4]
```diff
@@ -8,4 +8,8 @@
   &--error {
     border: 2px solid get-color(error);
   }
+
+  &.wmnds-col-1-2 {
+    width: 50%;
+  }
 }
```
Comment (daylesalmon) [inline src/wmnds/components/form-elements/input/_input.scss:15]: This isn't needed, just wrap the html element with a <div class="wmnds-col-1-2"><input /></div>.

File: src/wmnds/components/message/summary/_summary.njk [hunk 5]
```diff
@@ -18,23 +21,27 @@
...
     <h3 class="wmnds-msg-summary__title">{{ params.title }}</h3>
+    {% if params.subtitle %}<h3>{{ params.subtitle }}</h3>{% endif %}
   </div>
-
   <div class="wmnds-msg-summary__info">
     {{ params.message | safe }}
   </div>
 </div>
-
 {% endmacro %}
```
Comment (daylesalmon) [inline src/wmnds/components/message/summary/_summary.njk:47]: End of file needs a blank line.

File: src/wmnds/components/message/summary/_summary.njk [hunk 6]
```diff
@@ -7,6 +7,9 @@
 {# Success #}
 {% elseif params.type === "success-fill" %}
   {% set iconName = 'general-success' %}
+{# Success #}
+{% elseif params.type === "success-fill-no-icon" %}
```
Comment (daylesalmon) [inline src/wmnds/components/message/summary/_summary.njk:11]: Like this addition, can you add it to the https://deploy-preview-386--wmnetwork.netlify.app/components/message/ page.

File: src/wmnds/components/message/summary/_summary.njk [hunk 7]
```diff
@@ -7,6 +7,9 @@
...
+  {% set iconName = '' %}
 {# Warning #}
```
Comment (daylesalmon) [inline src/wmnds/components/message/summary/_summary.njk:13]: Can you check your whitespace on the component preview. For instance: https://deploy-preview-386--wmnetwork.netlify.app/components/message/ "success (fill) whitespace is a bit weird.

File: src/wmnds/components/message/summary/_summary.scss [hunk 8]
```diff
@@ -56,6 +56,17 @@ $wmnds-msg-icon-margin-right: $size-sm;
     }
   }
 
+  // Sucess fill no icon
```
Comment (daylesalmon) [inline src/wmnds/components/message/summary/_summary.scss:59]: Typo here, should be "success".

File: src/wmnds/components/table/_table.njk [hunk 9]
```diff
@@ -1,22 +1,38 @@
...
     {%- for j in range(1, params.rows + 1) %}
     <tr>
-      <th scope="row">Cell</th>
+      <th scope="row" class="{{params.cellClasses[0]}}">{% if params.values %}{{params.values[j-1][0]}}{% else %}Cell{% endif %}</th>
       {%- for k in range(1, params.cols) %}
-      <td>Cell</td>
+        <td class="{{params.cellClasses[k]}}">{% if params.values %}{{params.values[j-1][k] | safe }}{% else %}Cell{% endif %}</td>
       {%- endfor %}
     </tr>
     {%- endfor %}
```
Comment (daylesalmon) [inline src/wmnds/components/table/_table.njk:38]: Looks good, again just check the whitespace in the component preview code snippet.

File: src/wmnds/patterns/question/confirmation-message/_confirmation-message.njk [hunk 10]
```diff
@@ -0,0 +1,15 @@
...
+
+{{-
+    wmndsMsgSummary({
+        type: 'success-fill-no-icon',
+        title: 'Request complete',
+        subtitle: 'You can now start using your Swift card',
+        message: 'Your reference number is <b>12345</b>',
+        classes: 'wmnds-text-align-center'
+    }) | trim 
+-}}
```
Comment (daylesalmon) [inline src/wmnds/patterns/question/confirmation-message/_confirmation-message.njk:13]: Just check the indent in the code preview.

File: src/wmnds/patterns/question/reference/_reference.njk [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+        </div>
+
+    </div>
+    <div class="wmnds-col-1  wmnds-p-sm">
+        {{ params.button  | safe  | trim | indent(4) }}
+    </div>
+
+</div>
+
+{% endmacro %}
```
Comment (daylesalmon) [inline src/wmnds/patterns/question/reference/_reference.njk:47]: Check your indenting again.
Also, you are missing the form and fieldset elements like your "form element" pattern.

File: src/wmnds/patterns/question/summary/_summary.njk [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+        }) | trim | indent(4)
+    }} 
+    {%- for table in tables %}
+        {{ table | trim | indent(4) | safe }}
+    {%- endfor %}
+    {{ html | trim | indent(4) | safe }}
+    {{ params.button | trim | indent(4) | safe }}
+</div>
+
+{% endmacro %}
```
Comment (daylesalmon) [inline src/wmnds/patterns/question/summary/_summary.njk:25]: Indentation needs fixing in code snippet.

File: src/www/pages/patterns/question-form/index.njk [hunk 13]
```diff
@@ -0,0 +1,158 @@
...
+    <p>This is a confirmation message to show that the user has successfully submitted the data in the form, along with a reference or any information which the user needs to be made aware of.</p>
+    {% from "wmnds/patterns/question/confirmation-message/_confirmation-message.njk" import wmndsConfirmationMessage %}
+    {{
+        compExample([
+            wmndsConfirmationMessage() | indent(4) | trim
+        ])
+    }}
+    {# End Confirmation Message #}
+
+{% endblock %}
```
Comment (daylesalmon) [inline src/www/pages/patterns/question-form/index.njk:158]: New line needed at end of file.

File: src/www/pages/patterns/question-form/index.njk [hunk 14]
```diff
@@ -0,0 +1,158 @@
...
+                button: 
+                    wmnButton({
+                        classNames: "wmnds-btn--start",
+                        text: "Accept & Send ",
+                        iconRight: "general-chevron-right"
+                    }) 
+            }) 
+
+        ]) 
+    }}
```
Comment (daylesalmon) [inline src/www/pages/patterns/question-form/index.njk:142]: This could perhaps be tidied up with the caller method as mentioned by Gil.

**Review by daylesalmon — CHANGES_REQUESTED** [review]
Mainly indentation changes, but looking good so far!

**Review by gldgrnt — COMMENTED** [review]
I think Dayle mentioned all the things that I was going to say!

File: src/www/pages/patterns/question-form/index.njk [hunk 15]
```diff
@@ -0,0 +1,125 @@
...
+    {% from "wmnds/components/button/_button.njk" import wmnButton %}
+    {{
+        compExample([
+            wmndsSummary({
+                question: "Check your answers before reinstating your Direct Debit",
+                html:
+"<h3>Now send your request</h3>
+<p>By submitting this request you are confirming that, to the best of your knowledge, the details you are providing are correct.<br>
+<br>You consent to West Midlands Combined Authority possibly changing your Direct Debit date if it usually would occur in the next 10 days.<br>
+<br>You also consent to West Midlands Combined Authority reinstating your Direct Debit instruction.</p>",
```
Comment (daylesalmon) [inline src/www/pages/patterns/question-form/index.njk:99]: Can you just indent this so it looks a bit neater... 
I'm working on a separate branch for getting the indentation in code previews etc. to work correctly.

File: src/www/pages/styles/utility-classes/index.njk [hunk 16]
```diff
@@ -194,6 +210,21 @@ Where <code class="wmnds-website-inline-code">{size}</code> is one of:
...
 
+<hr>
+{# Background #}
+<h2>Background Colour</h2>
+<p>
+  Changes background to the intended colour.
+</p>
 
+<pre class="wmnds-col-1 wmnds-col-md-1-2">
+  <code class="scss" tabindex="0">
```
Comment (daylesalmon) [inline src/www/pages/styles/utility-classes/index.njk:221]:
```suggestion
  <code class="wmnds-show-more-ignore scss" tabindex="0">
```

Adding this class stops the "show more code" bottom from appearing.

I've also noticed this is happening on the "containers" code preview on this page. Can you also add the same class to the code preview too.

**Review by daylesalmon — CHANGES_REQUESTED** [review]
Almost ready to go.

**Review by gldgrnt — DISMISSED** [review]
Other than the changes Dayle suggested, it looks all good to me, good work on all the indentation stuff

**Review by gldgrnt — COMMENTED** [review]
> Other than the changes Dayle suggested, it looks all good to me, good work on all the indentation stuff

Same as this, but I didn't mean to approve it before !

File: src/www/pages/styles/utility-classes/index.njk [hunk 17]
```diff
@@ -194,6 +210,21 @@ Where <code class="wmnds-website-inline-code">{size}</code> is one of:
...
+  Changes background to the intended colour.
+</p>
 
+<pre class="wmnds-col-1 wmnds-col-md-1-2">
+  <code class="wmnds-show-more-ignore scss" tabindex="0">
+    .bg-default
+    .bg-white
+    .bg-grey
+    .bg-yellow
+    .bg-transparent
```
Comment (daylesalmon) [inline src/www/pages/styles/utility-classes/index.njk:226]: Can you update the bg names to the same colour names on the [colour palettes page](https://deploy-preview-386--wmnetwork.netlify.app/styles/colour-palettes/) so they are in sync.

An example is changing bg-yellow to bg-planned-disruption.

**Review by daylesalmon — CHANGES_REQUESTED** [review]
One last change and it should be ready to go live.

**Review by gldgrnt — APPROVED** [review]
Looks all good to me

**Review by daylesalmon — APPROVED** [review]

**Pull request merged**